### ISSUE TITLE
Changes to imported modules via FromFile are not updated?

### ISSUE THREAD

**Issue by cbsteh** [Roger-luo/FromFile.jl#30]
I'm using `FromFile` to import module `A` into module `B`. But if I change anything in module `A`, these changes are not seen by module `B`. For instance:

````
# A.jl
module A
varA = 1
end

# B.jl
module B
using FromFile
@from "A.jl" import A
@show A.varA   # outputs 1 regardless what happens in module A
end
````

If I was to change `varA = 100` in module `A`, `@show A.varA` in module `B` still outputs the old value of `1`. Changes to module `A` is only updated if I reload REPL. Is this by design?

**Comment by patrick-kidger** [comment]
How are you changing `varA = 100`?

**Comment by cbsteh** [comment]
The `varA` is changed by me directly editing `A.jl` file.

**Comment by patrick-kidger** [comment]
Are you using Revise.jl? i.e. why do you expect changes to a file to affect a running program anyway?

**Comment by cbsteh** [comment]
Yes, I was using Revise.jl, but when I noticed this issue, I commented out `using Revise` and reloaded REPL. The sample code I posted here is just an example to highlight my problem when changes to a module (that is still under development) are not seen.

**Comment by patrick-kidger** [comment]
If you've disabled Revise.jl then changes to a file will certainly never be seen. FromFile doesn't try to do any code-reloading; it's just like normal Julia in that regard.

Are you asking for compatibility between FromFile.jl and Revise.jl? That'd certainly be nice to have, but it's not clear from your question that that's what you're looking to discuss.

**Comment by Roger-luo** [comment]
I don't think Revise would track this at the moment, unless we generate `includet` instead I think, since it won't detect file A.jl as a dependency of B due to the absence of the standard `include` structure IIUC

**Comment by patrick-kidger** [comment]
Yep. (I tested this too just to be sure.) Revise will notice entirely new `@from` statements in the top level module, but won't respond to updates in any files that have been `@from`'d.

I'm not sure off the top of my head what the correct solution is. `includet` [seems to be treated](https://timholy.github.io/Revise.jl/stable/user_reference/#Revise.includet) as a "hack that works" (as opposed to having Revise track a package), so probably to be avoided in a library.

**Comment by Roger-luo** [comment]
> I'm not sure off the top of my head what the correct solution is. includet seems to be treated as a "hack that works" (as opposed to having Revise track a package), so probably to be avoided in a library.

Yes, but I would see this package as a workaround hack of Julia until 4600 get resolved too tbh. So I'm thinking maybe we can do this:

1. if we detect `Revise` has been loaded, we generate `includet`
2. if not we generate `include`

?

**Comment by patrick-kidger** [comment]
Hah! I've kind of given up on 4600 tbh. I personally think of FromFile as being in the same category as Revise or OhMyREPL -- non-stdlib but not far off essential. (The varying opinions on this in the community notwithstanding, of course.)

Anyway, what you suggest could be one way of doing it, hackiness notwithstanding.

Alternatively -- and I don't know the details of this -- I've noticed that
```julia
module MyPackage
include("blah.jl")
end
```
then
```julia
julia> using Revise
julia> using MyPackage
```
then edit the contents of `blah.jl`
then this will be picked up on in the REPL.

That is, Revise is smart enough to track includes in imported modules. Given that that's basically the situation we're in -- we cache a module that includes a file -- it's a little surprising to me that this doesn't work out-of-the-box. Probably we're just hitting some edge case in Revise.

**Comment by MariusDrulea** [comment]
Any progress on the topic? I also think having this functionality would help a lot.

**Comment by patrick-kidger** [comment]
I don't think either me or @Roger-luo are looking at it, no. If anyone gets to the bottom of it we'd be happy to accept a PR.

**Comment by MariusDrulea** [comment]
I will give it a try although my understanding of Julia metaprogramming is currently limited. I have identified two possible solutions to use Revise here. 
Option 1:
1. Checking if Revise is loaded: https://stackoverflow.com/questions/25575406/list-of-loaded-imported-packages-in-julia
2. change "include" to "includet" in line 39: file_module = Base.eval(root, :(module $(file_module_sym); include($path); end))

Option 2:
FromFile checks if Revise is installed and directly calls "using Revise" in the FromFile implementation

Possible problems: both solutions are probably a hack and not the wisest solutions as pointed out in the above comments

**Comment by Roger-luo** [comment]
I don't really have much time this month, but I might take a closer look at what's going on with Revise later

**Comment by MariusDrulea** [comment]
Status today: includet implemented, does not compile

I have added Revise as a dependency to FromFile and changed to includet instead of include. For some reason, Julia complains includet is not visible: `UndefVarError: includet not defined`
```
using Revise  # top level

function from_m(m::Module, s::LineNumberNode, path::String, root_ex::Expr)
           ................................................
           file_module = Base.eval(root, :(module $(file_module_sym); includet($path); end))
           ................................................
end
```

**Comment by Roger-luo** [comment]
you need to put full path in your expression? or just insert the `includet` object in the expression

**Comment by MariusDrulea** [comment]
@Roger-luo the path in that line is already processed to include the absolute path.

I had to include `using Revise` inside the `Base.eval` method and this way the compilation succeeds.

**Revise + include**: `file_module = Base.eval(root, :(module $(file_module_sym); using Revise; include($path); end))`
**Behaviour**: does not track changes

**Revise + includet**: `file_module = Base.eval(root, :(module $(file_module_sym); using Revise; includet($path); end))`
**Behaviour**: FromFile compiles, but when I run my code I get: `UndefVarError: MyModule not defined` for one of the modules I use.

Any further ideas?

**Comment by Roger-luo** [comment]
I think it would be easier to discuss If you submit a draft PR so that people can actually look at the code

> the path in that line is already processed to include the absolute path.

Sorry, I meant using the absolute module path via `GloablRef`

**Comment by MariusDrulea** [comment]
here is the PR with **Revise + includet**: https://github.com/Roger-luo/FromFile.jl/pull/31

**Comment by MariusDrulea** [comment]
> Sorry, I meant using the absolute module path via `GloablRef`

I'm a bit lost here. Do you mean something like this?
```
file_module = GlobalRef(root, file_module_sym) # --- instead of file_module = getfield(root, file_module_sym)
```
Then I get compilation error with fullname(file_module)...it's not defined for GlobalRefs

**Comment by MariusDrulea** [comment]
@Roger-luo, @patrick-kidger what if we invite somebody from Revise to have a look over the FromFile issue? It is faster for them to understand FromFile than for us to understand Revise. Is Tim Holy the right candidate for this?

**Comment by patrick-kidger** [comment]
If they're happy to do so then sure. It's certainly true that FromFile is much smaller than revise.

**Comment by MariusDrulea** [comment]
hm... I have created a julia package and did some tests. **Revise + FromFile works just fine.** In the past I was using include("some_file.jl") inside REPL, and Revise is not meant to work this way. Revise works best with packages. See here: https://discourse.julialang.org/t/basic-revise-jl-usage-when-developing-a-module/19140.

Directory structure:
```
MyPkg:
- src
- - MyPkg.jl
- - a.jl
- - b.jl
- Project.toml
- Manifest.toml
```
a.jl
```
module a
    export afun
    afun() = println("a")
end
```

b.jl
```
module b
    export bfun

    using FromFile
    @from "a.jl" using a

    function bfun()
        afun()
        println("b")
    end
end
```

MyPkg.jl
```
module MyPkg
    export fun

    using FromFile
    @from "b.jl" using b

    function fun()
        bfun()
    end

end # module
```

Now in the REPL I enter the following:
```
activate .
using Revise
using MyPkg

fun() # here I get "a b" printed. 
```

Now edit afun() to print a3, and re-call fun() in REPL:
`fun() # prints"a3 b", as expected`

**Comment by MariusDrulea** [comment]
Now the REPL mode:
```
julia> include("MyPkg.jl")
Main.MyPkg

julia> Main.MyPkg.fun()
a
b

julia> println("Now edit the afun() in a.jl to print a3. Then re-include the file.")
Now edit the afun() in a.jl to print a3. Then re-include the file.

julia> include("MyPkg.jl")
WARNING: replacing module MyPkg.
Main.MyPkg

julia> Main.MyPkg.fun()
a
b

julia> println("this is not what we want, we should have a3 b.")
this is not what we want, we should have a3 b

```

If I use the standard include("a.jl") and include("b.jl") instead of @from, I get the expected results.

**Comment by MariusDrulea** [comment]
Possible solution for the REPL mode:
If `root === Main` we are in the REPL mode, right? In this case, we shall Base.eval(...) even if the file_module_sym is already defined.

```
    if root === Main
        file_module_sym = Symbol(path)
    else
        file_module_sym = Symbol(relpath(path, pathof(root)))
    end

    if isdefined(root, file_module_sym) ------> && !(root === Main) # <---------------------
        file_module = getfield(root, file_module_sym)
    else
        file_module = Base.eval(root, :(module $(file_module_sym); include($path); end))
    end
```

**Comment by Roger-luo** [comment]
I believe it should be 

    __module__ == Main && isinteractive()

Main doesn't necessarily indicate the REPL usage

**Comment by MariusDrulea** [comment]
I have implemented the feature, it works, but if I have a file "a.jl" included in several other files, the module "a" is replaced for every include.  We shall optimize to "replace" a module only once per call of `include("tutorial\\no_pkg\\d.jl")`. Any suggestions?

[a.jl.txt](https://github.com/Roger-luo/FromFile.jl/files/7423873/a.jl.txt)
[b.jl.txt](https://github.com/Roger-luo/FromFile.jl/files/7423874/b.jl.txt)
[c.jl.txt](https://github.com/Roger-luo/FromFile.jl/files/7423875/c.jl.txt)
[d.jl.txt](https://github.com/Roger-luo/FromFile.jl/files/7423876/d.jl.txt)


```
julia> include("tutorial\\no_pkg\\d.jl")
WARNING: replacing module d.
WARNING: replacing module D:\Software\repos\grid_scripts\julia\tutorial\no_pkg\a.jl.
WARNING: replacing module D:\Software\repos\grid_scripts\julia\tutorial\no_pkg\b.jl.
WARNING: replacing module D:\Software\repos\grid_scripts\julia\tutorial\no_pkg\a.jl.
WARNING: replacing module D:\Software\repos\grid_scripts\julia\tutorial\no_pkg\c.jl.
WARNING: replacing module D:\Software\repos\grid_scripts\julia\tutorial\no_pkg\a.jl.
WARNING: replacing module D:\Software\repos\grid_scripts\julia\tutorial\no_pkg\b.jl.
WARNING: replacing module D:\Software\repos\grid_scripts\julia\tutorial\no_pkg\a.jl.
```

**Comment by jondeuce** [comment]
Hi all, love the concept of FromFile.jl. I've been looking into this issue a bit, since I would really like to seriously try FromFile.jl but I am also pretty attached to the niceties of Revise.jl.

Without knowing too much about the Revise internals, the issue seems to be coming from the fact that Revise can track either a) `using` and `import` statements for packages on the `LOAD_PATH` via hooks into Base internals, or b) source files which it is manually told to track via `Revise.includet` or `Revise.track`.

In particular, as far as I can tell Revise will not just track any old module. This is a problem for FromFile, since you are [generating `module X; include(path); end` statements](https://github.com/Roger-luo/FromFile.jl/blob/816479581cda263735d620f3494f4a8984c04383/src/FromFile.jl#L43) in a given `root` module, and Revise won't then automatically track `root.X`.

For example, if one does the following:

```julia
# A.jl
module A
a() = 1
end

julia> module Awrap
       include("src/A.jl")
       end
julia> using .Awrap
```

Then changes `A.jl` are not tracked by Revise, and e.g. the `Awrap.A.a` method cannot be updated by editing `A.jl`.

So, without changes to Revise, I see two options to fix this:
1. (This seems very sketchy to me, and I'm not really seriously suggesting it) Generate files for mini packages on the fly and mess with `LOAD_PATH` so that they can be discovered and loaded with `using` or `import`, which Revise knows how to handle
2. Explicitly tell Revise to track the module generated by FromFile.

For the second option, It's actually very easy to hack up a temporary solution using `Revise.track`: if [in the else branch of this `if` statement](https://github.com/Roger-luo/FromFile.jl/blob/816479581cda263735d620f3494f4a8984c04383/src/FromFile.jl#L44) you add something like

```julia
if root == Main && isinteractive() && isdefined(Main, :Revise)
    Main.Revise.track(file_module, path)
end
```

then changes to the generated module `file_module` will track changes to the code at `path`. Per the [Revise docs](https://timholy.github.io/Revise.jl/stable/user_reference/#Revise.includet), this is similar to how `includet` works internally.

Probably there is a way to do this more cleanly (without depending on Revise directly) using [Requires.jl](https://github.com/JuliaPackaging/Requires.jl), which I think could be very nice. If this is a direction you all would be interested in, let me know and I can try to put together a PR.

UPDATE: [I implemented a quick `@requires`-based solution as a proof-of-concept](https://github.com/jondeuce/FromFile.jl). Seems to work great, but haven't added any tests yet.

**Comment by patrick-kidger** [comment]
This looks pretty cool!

We'd definitely like to have better compatibility here. If you can put something together that is robust then that'd be great.

I think there's also funny edge cases to be careful of here,  e.g. what is someone defines a variable in `Main` that just happens to be called `Revise`? And should we be tracking / does Revise track changes to packages, i.e. without `Main` as the root? (E.g. packages installed as `dev`.)

But modulo those concerns what you've done looks pretty convincing (to me, at least, as a non-Revise-user). @Roger-luo WDYT?

**Comment by jondeuce** [comment]
Great! And yes I agree re: your concerns about the funny edge cases; I implemented a `@requires`-based solution instead that is more robust (files will only be tracked if Revise has already been loaded in an interactive session).

Re: tracking changes to packages without `Main` as the root: [Revise can track changes to a rather impressive variety of code](https://timholy.github.io/Revise.jl/stable/#What-Revise-can-track-1) (`dev`-ed packages, `Base` methods, etc.), so I don't think that should be an issue. Re: whether files included using `@from` _should_ be tracked: I think that it's fair to assume that if a user has loaded Revise in an interactive session, then they probably want all possible code changes tracked.

I'll make a quick PR so that we have something concrete to discuss.

**Comment by Roger-luo** [comment]
should be closed by #36 now
